# Project Description
The coronavirus took the entire world by surprise, changing everyone's daily routine. City dwellers no longer spent their free time outside, going to cafes and malls; more people were home, reading books. That attracted the attention of startups that rushed to develop new apps for book lovers.
You've been given a database of one of the services competing in this market. It contains data on books, publishers, authors, and customer ratings and reviews of books. This information will be used to generate a value proposition for a new product.

# Purpose Project

The project aims to utilize the provided database to create a value proposition for a new product or service for book lovers. By analyzing the data, trends, and preferences of readers, opportunities will be identified to develop a unique offering. The goal is to create a compelling proposition that differentiates the new product from existing services and meets the needs of book enthusiasts.

# Connecting to the Database

In [2]:
# import libraries
import pandas as pd
from sqlalchemy import create_engine

In [3]:
db_config = {'user': 'praktikum_student',         # username
             'pwd': 'Sdf4$2;d-d30pp', # password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # connection port
             'db': 'data-analyst-final-project-db'}          # the name of the database

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [4]:
books = 'SELECT * FROM books'
books = pd.read_sql(books, engine)

authors = 'SELECT * FROM authors'
authors = pd.read_sql(authors, engine)

publishers = 'SELECT * FROM publishers'
publishers = pd.read_sql(publishers, engine)

ratings = 'SELECT * FROM ratings'
ratings = pd.read_sql(ratings, engine)

reviews = 'SELECT * FROM reviews'
reviews = pd.read_sql(reviews, engine)

# Review Data

In [5]:
books.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [6]:
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


In [7]:
books.duplicated().sum()

0

In [8]:
authors.head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [9]:
authors.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


In [10]:
authors.duplicated().sum()

0

In [11]:
publishers.head()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [12]:
publishers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


In [13]:
publishers.duplicated().sum()

0

In [14]:
ratings.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [15]:
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


In [16]:
ratings.duplicated().sum()

0

In [17]:
reviews.head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [18]:
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


In [19]:
reviews.duplicated().sum()

0

**Observation**

There are 5 seperated data. The data contains the the information of books, the authors, the publishers, the ratings, and lastly the reviews. From all these datas, lets create a compelling proposition that differentiates the new product from existing services and meets the needs of book enthusiasts.

# Number of Books

Find the number of books released after January 1, 2000.

In [20]:
books_query = '''
SELECT COUNT(*) AS num_books
FROM books
WHERE publication_date > '2000-01-01'
'''

# Execute the SQL query using pandas and the established database connection
books_query = pd.read_sql(books_query, engine)
books_query

num_books
0        819

There are 819 books in the database that released from 2000 until the latest date. 

# Users Review and Avg Rating

Find the number of user reviews and the average rating for each book.

In [21]:
# SQL query to count the number of user reviews and calculate the average rating for each book
review_rating = '''
SELECT r.book_id, COUNT(DISTINCT r.review_id) AS num_reviews, AVG(rt.rating) AS avg_rating
FROM reviews AS r
JOIN ratings AS rt ON r.book_id = rt.book_id
GROUP BY r.book_id
ORDER BY num_reviews DESC;
'''

# Execute the SQL query using pandas and the established database connection
review_rating = pd.read_sql(review_rating, engine)

# Display the result
review_rating

book_id  num_reviews  avg_rating
0        948            7    3.662500
1        854            6    3.772727
2        656            6    4.264151
3        734            6    4.206897
4        963            6    3.977273
..       ...          ...         ...
989      465            1    3.500000
990      446            1    3.000000
991      431            1    4.000000
992       92            1    3.500000
993      413            1    4.000000

[994 rows x 3 columns]

There are 994 books that have been saved into the database. However, from the books ID the number not matches. it means that there have a missing books from the database. The table have been calculated the number of reviews and the average rating for each books in the database.

# Greatest publisher

Identify the publisher that has released the greatest number of books with more than 50 pages (this will help you exclude - brochures and similar publications from your analysis).

In [22]:
# SQL query to identify the publisher with the greatest number of books with more than 50 pages
publisher_query = '''
SELECT p.publisher, COUNT(b.book_id) AS num_books
FROM books AS b
JOIN publishers AS p ON b.publisher_id = p.publisher_id
WHERE b.num_pages > 50
GROUP BY p.publisher
ORDER BY num_books DESC
LIMIT 1
'''

# Execute the SQL query using pandas and the established database connection
publisher_query = pd.read_sql(publisher_query, engine)

publisher_query

publisher  num_books
0  Penguin Books         42

The greatest publisher in the database with has the greatest number of books with more than 50 pages are Penguin Books with 42 books. 

 # Greatest Author

Identify the author with the highest average book rating: look only at books with at least 50 ratings.

In [23]:
# SQL query to identify the author with the highest average book rating for books with at least 50 ratings
author_query = '''
SELECT a.author, AVG(rating) AS avg_rating
FROM authors AS a
JOIN (
    SELECT b.author_id, r.rating
    FROM books AS b
    JOIN ratings AS r ON b.book_id = r.book_id
    GROUP BY b.author_id, r.rating
    HAVING COUNT(*) >= 50
) AS ra ON a.author_id = ra.author_id
GROUP BY a.author
ORDER BY avg_rating DESC

'''

# Execute the SQL query using pandas and the established database connection
author_query = pd.read_sql(author_query, engine)
author_query

author  avg_rating
0              J.R.R. Tolkien         4.5
1  J.K. Rowling/Mary GrandPré         4.0
2             Stephenie Meyer         3.5

The author that have the highest average book rating (for books with at least 50 ratings) are J.R.R. Tolkien. The average rating are 4.5.

# Avg Text Review

Find the average number of text reviews among users who rated more than 50 books.

In [24]:
# SQL query to find the average number of text reviews among users who rated more than 50 books
users_query = '''
SELECT AVG(review_count) AS average_review_count
FROM (
    SELECT COUNT(*) AS review_count
    FROM reviews
    WHERE username IN (
        SELECT username
        FROM ratings
        GROUP BY username
        HAVING COUNT(DISTINCT book_id) > 50
    )
    GROUP BY username
) AS subquery;
'''

# Execute the SQL query using pandas and the established database connection
users_query = pd.read_sql(users_query, engine)
users_query

average_review_count
0             24.333333

The results show the users who have rated more than 50 books, the average number of text reviews per user is approximately 24.333333. This indicates that, on average, each user in that group has written around 24 text reviews.